Hash Data Structure

jabolina opened the pull request:
Adding the initial design draft for the Hash data structure. I think there are some unknown unknowns for me here, but I tried to cover at least a bit of everything.

jabolina committed:
Adding Hash data structure design

karesti commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,174 @@
+[source,java]
+.Hash.java
+----
+public interface Hash<K, V> { // <1>
```
How is this different from cache ?

tristantarrant replied:
Because it is stored as a single entry in a cache.

karesti replied:
ok, now I get it

wburns replied:
Looking at this, it looks awfully close our MultiMap, just instead of a Collection it is a Map. Could we not build on top of that?

jabolina replied:
We could build upon MultiMap without duplicates, storing key-value pairs. But I am unsure if we can compare for equality based only on the key. On the other hand, maybe we could abstract some of the MultiMap implementations which work over a Collection to work on another structure instead of creating everything from scratch. I would need to look at that, though.

pruivo replied:
isn't it the same as `FineGrainedAtomicMap` that we dropped and replaced with the `Grouping API`?

tristantarrant commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,174 @@
+----
+
+<1> The Hash data structure is a generic interface over `K` and `V`. Should we extend `Map<K, V>`?
+This way we could provide a wide range of operations over the Hash.
```
I would avoid extending `Map` as that has a million methods that have made it a pain in the past for our `Cache`.

karesti replied:
yes, why not using map directly if we do that.

tristantarrant replied:
Some further information:
* `Map` is a sync-only API, so this does not help with the async and mutiny variants
* Some of the `Map` methods have effects, such as returning the previous value on `put` and `remove` which are costly in remote scenarios.

karesti commented on Cache-Hash-Data-Structure.adoc:
````diff
@@ -0,0 +1,174 @@
+To create and use a `HashCache<?>`:
+
+```java
+      EmbeddedCacheManager cm = ...
````
new api ?

karesti commented on Cache-Hash-Data-Structure.adoc:
````diff
@@ -0,0 +1,174 @@
+
+```java
+      EmbeddedCacheManager cm = ...
+      HashCacheManager hashCacheManager = EmbeddedHashCacheManagerFactory.from(cm);
````
this is the legacy API, we need this in the new API only

tristantarrant replied:
Let's not implement anything on the legacy API please.

rigazilla commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,174 @@
+
+A `Hash<?, ?>` structure is local to a node. Meaning that operations applied to the local copy must be submitted
+through the public API before taking effect. After that, the `Hash<?, ?>` is distributed/replicated according to
+the configuration.
```
How is the intended use case?
The user operates on an Hash object and then uses an HashCache obj to submit the changes?

tristantarrant replied:
I would say that individual operations are applied immediately to the underlying storage (remote cache). When using tx/batching, these would be collected and sent in a single operation.

jabolina replied:
> The user operates on an Hash object and then uses an HashCache obj to submit the changes?

Yeah, I was thinking of something of that sort. I'll update the documentation to include single operations without transferring the whole object. For multiple operations, I was thinking of keeping track of the local changes, and once the user submits the changes to the server, we merge the local Hash with the server's Hash.

I'll update the docs.

jabolina commented:
Thanks, everyone, for the comments. I updated the document with all the suggestions. While at it, I described more about how we handle the operations. A point of attention is how we should handle operations that directly reference a Hash, for example, `HashCache.put("hash", "property", "value")`, I added the two options I see for that in the document.

rigazilla commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+----
+
+<1> Returning a `Map<HK, HV>` to contain all the vargs keys.
+<2> The `submit` method is used to apply the local changes to the server's Hash. The server's `Hash<?, ?>` is returned.
```
Do we really need the return value? Hash could be very big. Returning on request or only if submit fails?

jabolina replied:
Yeah, but not strictly. I'll add some void methods, too. In one case, for example, the RESP connector needs to return what has changed. But I was aiming at something more general instead of focusing only on the RESP needs.

rigazilla commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+<1> The Hash data structure is a generic interface over `K` and `V`. We do not extend `Map<K, V>`, instead we add methods
+as needed.
+
+A `Hash<?, ?>` structure is local to a node. Meaning that operations applied to the local copy must be submitted
```
I feel a disturbance about Hash. I mean, if it's detached from cache isn't it just a Map do we need a new type? On the other side if individual operations are submitted immediately then some methods seems to overlap with HashCache

wburns replied:
I agree, I did not like the naming on the RESP commands either. A Hash to me is not a Map structure but rather a numeric derived from some other fields.

jabolina replied:
To be completely honest, even after writing the word "Hash" so many times, it still sounds weird as a structure name. I thought the same, i.e., a hash is a value after applying a hashing function (or even the function)... Regarding naming, I would not mind changing it to Map(ping), Document, Structure, Container, or something else.

And as the structure Hash, which the user interacts with, would help to keep track of the operations. For example, if we return a Map and the user removes one of the keys, we wouldn't notice that when merging with the server's content. Or, to better reflect, rename the structure HashView, HashCopy, or something of that sort.

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+----
+public interface HashCache<K> {
+
+   <HK, HV> HV put(K key, HK hashKey, HV hashValue);
```
We should probably name this method `getAndPut` and have just a `put` method that has no returned.

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+[source, java]
+.HashCache.java
+----
+public interface HashCache<K> {
```
~~This entire interface is blocking, we probably need NonBlocking versions first.~~

Reading further it looks like we would have all 3 of the variants 👍

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+
+   <HK, HV> HV put(K key, HK hashKey, HV hashValue);
+
+   <HK, HV> Hash<HK, HV> put(K key, Map<HK, HV> values);
```
Same here, we really need just a `put` and `getAndPut`.

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+
+   <HK> boolean containsKey(K key, HK hashKey);
+
+   <HK, HV> Hash<HK, HV> submit(Hash<HK, HV> hash); // <2>
```
Confused as to what this method actually does. Why is there no key tied to the hash?

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+
+   boolean remove(K key);
+
+   <HK> boolean remove(K key, HK hashKey, HK ... others);
```
A single boolean doesn't work here as some may be removed and some may not.

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+* CRUD hashes (providing a configuration to create a hash, templates not supported)
+* Display Hash content entries
+* Get values by key
+* In Indexing is enabled, hashes support query
```
How would it do this? Are we planning on having the hashes have their own key and value schema definitions? I don't see that mentioned.

jabolina replied:
This one is straight up a stretch, and one thing I had doubts about (the last bullet point at the end of the doc). If feasible, we would have something similar to the [JSON support](https://redis.io/docs/stack/json/). Should I mark it optional? Or remove?

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+* Enable/Disable rebalancing
+* Availability get/set
+* Replace not supported
+* Listeners support to enable interactions with near-caching
```
Listeners are going to be very very very expensive for this without some sort of delta notification, which we do not have for listeners. Also listener API was never solidified for the new API sadly :(

jabolina replied:
I'll mark it optional. When we establish the listener API, we can come back to this.

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+   HashCache<String> users = hashes.get("users");
+
+   users.put("42", Map.of("auth", "token", "role", "user"));
+   System.out.println(hashCache.get("42")); // {"auth": "token", "role": "user"}
```
Is `hashCache` here supposed to be `users`?

Also from this code, I assume that an updates to the returned `HashCache` instance are immediately updated remotely and applied to the `Hashes` actual contents? We will need to make sure the interface is documented that that is done that way.

Only the `Hash` interface is the one that isn't propagated correct?

jabolina replied:
> Is `hashCache` here supposed to be `users`?

Yes, I'll fix it.


> Also from this code, I assume that an updates to the returned `HashCache` instance are immediately updated remotely and applied to the `Hashes` actual contents? We will need to make sure the interface is documented that that is done that way.
> 
> Only the `Hash` interface is the one that isn't propagated correct?

Correct. Only the operations directly over the `HashCache` takes place immediately. If operating over a `Hash` it needs to be sent to the `HashCache` to take effect.

I've been thinking more about this and I'll do an update to the `HashCache` API.

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+   Hashes hashes = ispn.sync().hashes();
+   HashCache<String> users = hashes.get("users");
+
+   Hash<String, String> hash = users.get("42");
```
So the `Hash` returned will always be not null I assume?

wburns replied:
This makes me wonder if we should have some sort of putHashIfAbsent type of operation, where you can do an initial request to populate the Hash without having to do an initial get round trip.

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+as needed.
+
+A `Hash<?, ?>` structure is local to a node. Meaning that operations applied to the local copy must be submitted
+through the public API before taking effect. After that, the `Hash<?, ?>` is distributed/replicated according to
```
I am debating about this. It seems we could also implement this in a slightly different fashion where the key is just a name of a cache and the underlying Hash is just a normal Cache which itself can be distributed etc. This would then give us delta changes between keys and we can use all of the existing features of Caches. I am not quite sold yet, especially as I have no idea what the query API would look like or listeners.

wburns commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+   hash.put("auth", "token");
+   hash.put("role", "user");
+
+   users.submit(hash);
```
So will this completely overwrite the hash if it is already exists or only apply the new values? It isn't clear or doesn't seem defined yet as to what this would do. I am guessing it completely overwrites those values, in which case you would lose any concurrent writes without a respective remove.

jabolina replied:
For this one, I was thinking to implement with compute and merge the values. We keep track of the operations with the `Hash`, but effectively all the local changes take precedence. I'll do an update in the docs and API.

jabolina commented:
@rigazilla and @wburns, thanks for the comments! I am going to update the doc and the `HashCache` API introducing your suggestions. I'll try to document better the intents with the API, evidencing the `Hash`needs to be sent to take effect. And rename that `submit` method to something meaningful.

pruivo commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+----
+public interface Hash<K, V> { // <1>
+
+   String getName();
```
what's the name? don't see any reference to it.

pruivo commented on Cache-Hash-Data-Structure.adoc:
```diff
@@ -0,0 +1,209 @@
+through the public API before taking effect. After that, the `Hash<?, ?>` is distributed/replicated according to
+the configuration.
+
+We follow a state-based CRDT approach, merging the local Hash with the server's Hash. Since the `Hash<?, ?>` structures
```
how are we going to merge when network partitions happen? Are we going to support cross-site?
